portfolio/src/main/java/com/google/sps/servlets/DataServlet.java

```diff
@@ -39,27 +40,42 @@
   
   @Override
   public void doGet(HttpServletRequest request, HttpServletResponse response) throws IOException {
-    Query query = new Query("Comment").addSort("timestamp", SortDirection.DESCENDING);
+    String defaultCommentLimit = "5";
+    int commentLimit = Integer.parseInt(getParameter(request, "numShown", defaultCommentLimit));
+
+    String defaultPageNumber = "0";
+    int pageNumber = Integer.parseInt(getParameter(request, "pageNumber", defaultPageNumber));
 
+    String defaultSortMethod = "DESCENDING";
+    String sortMethod = getParameter(request, "sortMethod", defaultSortMethod);
+
+    Query query = new Query("Comment");
+    
+    if (sortMethod.equals(defaultSortMethod)) {
+      query.addSort("timestamp", SortDirection.DESCENDING);
+    } else {
+      query.addSort("timestamp", SortDirection.ASCENDING);
+    }
     DatastoreService datastore = DatastoreServiceFactory.getDatastoreService();
     PreparedQuery results = datastore.prepare(query);
+    
 
     ArrayList<Comment> comments = new ArrayList<>();
```

Comment:
Unless you plan on using specific ArrayList methods, just declare this as the generic type List\<Comment\> comments = new ArrayList\<\>(); it's simpler and more maintainable to just treat objects as a small set of common generic types when possible and not have to worry about the specific implementation.